File: app/src/main/java/com/roostermornings/android/activity/DiscoverFragmentActivity.java
/*
 * Rooster Mornings Android.
 * Copyright (c)  2017 Roosta Media. All rights reserved.
 */

package com.roostermornings.android.activity;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.roostermornings.android.BaseApplication;
import com.roostermornings.android.R;
import com.roostermornings.android.activity.base.BaseActivity;
import com.roostermornings.android.adapter.DiscoverListAdapter;
import com.roostermornings.android.channels.ChannelManager;
import com.roostermornings.android.firebase.FA;
import com.roostermornings.android.dagger.RoosterApplicationComponent;
import com.roostermornings.android.domain.Channel;
import com.roostermornings.android.domain.ChannelRooster;
import com.roostermornings.android.sqlutil.AudioTableManager;
import com.roostermornings.android.sqlutil.DeviceAlarmTableManager;
import com.roostermornings.android.geolocation.GeoHashUtils;
import com.roostermornings.android.util.JSONPersistence;
import com.roostermornings.android.util.Toaster;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import butterknife.BindView;

public class DiscoverFragmentActivity extends BaseActivity implements DiscoverListAdapter.DiscoverAudioSampleInterface {

    @BindView(R.id.discoverListView)
    RecyclerView mRecyclerView;
    @BindView(R.id.toolbar_title)
    TextView toolbarTitle;
    @BindView(R.id.button_bar)
    LinearLayout buttonBarLayout;
    @BindView(R.id.swiperefresh)
    SwipeRefreshLayout swipeRefreshLayout;

    public static final String TAG = DiscoverFragmentActivity.class.getSimpleName();

    private ArrayList<ChannelRooster> channelRoosters = new ArrayList<>();

    private MediaPlayer mediaPlayer = new MediaPlayer();
    Future oneInstanceTaskFuture;

    private ChannelManager channelManager = new ChannelManager(this,this);

    private RecyclerView.Adapter mAdapter;

    @Inject BaseApplication AppContext;
    @Inject SharedPreferences sharedPreferences;
    @Inject JSONPersistence jsonPersistence;

    @Override
    protected void inject(RoosterApplicationComponent component) {
        component.inject(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initialize(R.layout.activity_discover);

        inject(BaseApplication.getRoosterApplicationComponent());

        //Notify user of no internet connection
        checkInternetConnection();

        //Final context to be used in threads
        final Context context = this;

        //UI setup thread
        new Thread() {
            @Override
            public void run() {
                //Setup day/night theme selection (based on settings, and time)
                setDayNightTheme();
                //Set highlighting of button bar
                setButtonBarSelection();
                //Set toolbar title
                Toolbar toolbar = setupToolbar(toolbarTitle, getString(R.string.discover));
                if(toolbar != null) {
                    toolbar.setNavigationIcon(R.drawable.md_nav_back);
                    toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            startHomeActivity();
                        }
                    });
                }

                // use this setting to improve performance if you know that changes
                // in content do not change the layout size of the RecyclerView
                mRecyclerView.setHasFixedSize(true);
                //Set up adapter for monitoring channelrooster objects
                mAdapter = new DiscoverListAdapter(channelRoosters, DiscoverFragmentActivity.this);
                //Use a linear layout manager
                RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(context);
                mRecyclerView.setLayoutManager(mLayoutManager);
                mRecyclerView.setAdapter(mAdapter);

                //Check if uses explor user prop has been set, and if not then create shared pref with default no
                String sharedPrefUsesExplore = sharedPreferences.getString(
                        FA.UserProp.uses_explore.shared_pref_uses_explore, "null");
                if(!sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.no)
                && !sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.started_explore_playback)
                && !sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.completed_explore_playback)) {
                    sharedPreferences.edit().putString(FA.UserProp.uses_explore.shared_pref_uses_explore,
                            FA.UserProp.uses_explore.no).apply();
                    FA.SetUserProp(FA.UserProp.uses_explore.class, FA.UserProp.uses_explore.no);
                }
            }
        }.run();

        if(!jsonPersistence.getChannelRoosters().isEmpty()) {
            channelRoosters.addAll(jsonPersistence.getChannelRoosters());
            mAdapter.notifyDataSetChanged();
        } else if(checkInternetConnection()) {
            if(!swipeRefreshLayout.isRefreshing()) swipeRefreshLayout.setRefreshing(true);
        }

        ChannelManager.Companion.setOnFlagChannelManagerDataListener(new ChannelManager.Companion.OnFlagChannelManagerDataListener() {
            @Override
            public void onChannelRoosterDataChanged(@NotNull ArrayList<ChannelRooster> freshChannelRoosters) {
                channelRoosters.clear();
                channelRoosters.addAll(freshChannelRoosters);
            }

            @Override
            public void onSyncFinished() {
                if(mAdapter != null) mAdapter.notifyDataSetChanged();
                swipeRefreshLayout.setRefreshing(false);
            }
        });

        /*
        * Sets up a SwipeRefreshLayout.OnRefreshListener that is invoked when the user
        * performs a swipe-to-refresh gesture.
        */
        swipeRefreshLayout.setOnRefreshListener(
                new SwipeRefreshLayout.OnRefreshListener() {
                    @Override
                    public void onRefresh() {
                        // This method performs the actual data-refresh operation.
                        // The method calls setRefreshing(false) when it's finished.
                        //Reload adapter data and set message status, set listener for new data
                        onPause();
                        channelManager.refreshChannelData(channelRoosters);
                    }
                }
        );

        //Set volume rocker to alarm stream
        setVolumeControlStream(AudioManager.STREAM_MUSIC);

        //Fetch content from Firebase
        channelManager.refreshChannelData(channelRoosters);
    }

    @Override
    public void onStart() {
        super.onStart();
        //Display notifications
        updateRoosterNotification();
        updateRequestNotification();
    }

    @Override
    public void onPause() {
        super.onPause();

        if(mediaPlayer != null) {
            mediaPlayer.release();
        }
        if(oneInstanceTaskFuture != null) oneInstanceTaskFuture.cancel(true);
        clearChannelRoosterMediaChecks();
        mAdapter.notifyDataSetChanged();

        //Persist channel roosters for seamless loading
        jsonPersistence.setChannelRoosters(channelRoosters);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            String query = intent.getStringExtra(SearchManager.QUERY);
            //use the query to search your data somehow
            handleSearch(query);
        }
    }

    private void handleSearch(String query) {
        ((DiscoverListAdapter)mAdapter).refreshAll(channelRoosters);
        ((DiscoverListAdapter)mAdapter).getFilter().filter(query);
    }

    public void notifyAdapter() {
        ((DiscoverListAdapter)mAdapter).refreshAll(channelRoosters);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_explore, menu);

        // Associate searchable configuration with the SearchView
        SearchManager searchManager =
                (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView =
                (SearchView) menu.findItem(R.id.search_explore).getActionView();
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(getComponentName()));

        //When searchView is closed, refresh data
        searchView.addOnAttachStateChangeListener(new View.OnAttachStateChangeListener() {

            @Override
            public void onViewDetachedFromWindow(View arg0) {
                // search was detached/closed
                notifyAdapter();
            }

            @Override
            public void onViewAttachedToWindow(View arg0) {
                // search was opened
            }
        });

        return true;
    }

    private void notifyDataSetChangedFromUIThread() {
        this.runOnUiThread(
                new Runnable() {
                    @Override
                    public void run() {
                        mAdapter.notifyDataSetChanged();
                    }
                }
        );
    }

    public void streamChannelRooster(final ChannelRooster channelRooster) {
        //The oneInstanceTask creates a new thread, only if one doesn't exist already
        ExecutorService executor = new ThreadPoolExecutor(1, 1, 1, TimeUnit.MINUTES,
                    new SynchronousQueue<Runnable>(),
                    new ThreadPoolExecutor.DiscardPolicy());

        //Check for an active internet connection - displays toast if none
        if(!checkInternetConnection()) {
            mediaPlayer.release();
            return;
        }
        //Do not proceed if active download in progress on current channelrooster
        if(channelRooster.isDownloading()) return;

        class oneInstanceTask implements Runnable {
            public void run() {
                StorageReference mStorageRef = FirebaseStorage.getInstance().getReference();
                final StorageReference audioFileRef = mStorageRef.getStorage().getReferenceFromUrl(channelRooster.getAudio_file_url());

                channelRooster.setDownloading(true);
                channelRooster.setPlaying(false);
                notifyDataSetChangedFromUIThread();

                mediaPlayer = new MediaPlayer();

                audioFileRef.getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>()
                {
                    @Override
                    public void onSuccess(Uri downloadUrl)
                    {
                        //The asynchronous nature of this task meant that even after reset, the download task would return and start the mediaplayer...
                        //Now by releasing the mediaplayer and initialising it before the async call we avoid this
                        if(mediaPlayer == null) return;
                        try {
                            mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
                            mediaPlayer.setDataSource(downloadUrl.toString());
                        } catch (IOException e) {
                            e.printStackTrace();
                            return;
                        } catch (IllegalStateException e) {
                            e.printStackTrace();
                            return;
                        }
                        mediaPlayer.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
                            @Override
                            public void onPrepared(MediaPlayer mp) {
                                mediaPlayer.start();

                                clearChannelRoosterMediaChecks();
                                channelRooster.setDownloading(false);
                                channelRooster.setPlaying(true);
                                notifyDataSetChangedFromUIThread();

                                //Set Firebase user prop for uses_explore
                                String sharedPrefUsesExplore = sharedPreferences.getString(
                                        FA.UserProp.uses_explore.shared_pref_uses_explore, "no");
                                if(sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.no)) {
                                    sharedPreferences.edit().putString(FA.UserProp.uses_explore.shared_pref_uses_explore,
                                            FA.UserProp.uses_explore.started_explore_playback).apply();
                                    FA.SetUserProp(FA.UserProp.uses_explore.class, FA.UserProp.uses_explore.started_explore_playback);
                                }

                                mediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
                                    @Override
                                    public void onCompletion(MediaPlayer mp) {
                                        clearChannelRoosterMediaChecks();
                                        notifyDataSetChangedFromUIThread();

                                        //Set Firebase user prop for uses_explore
                                        FA.SetUserProp(FA.UserProp.uses_explore.class, FA.UserProp.uses_explore.completed_explore_playback);
                                        sharedPreferences.edit().putString(FA.UserProp.uses_explore.shared_pref_uses_explore,
                                                FA.UserProp.uses_explore.completed_explore_playback).apply();
                                    }
                                });
                            }
                        });

                        mediaPlayer.setOnErrorListener(new MediaPlayer.OnErrorListener() {
                            @Override
                            public boolean onError(MediaPlayer mediaPlayer, int what, int extra) {
                                Toaster.makeToast(AppContext, "Content streaming failed.", Toast.LENGTH_SHORT).checkTastyToast();
                                clearChannelRoosterMediaChecks();
                                notifyDataSetChangedFromUIThread();
                                return true;
                            }
                        });

                        mediaPlayer.prepareAsync();
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception exception) {
                        // Handle any errors
                        Toaster.makeToast(AppContext, "Content streaming failed.", Toast.LENGTH_SHORT).checkTastyToast();
                        clearChannelRoosterMediaChecks();
                        mAdapter.notifyDataSetChanged();
                    }
                });
            }
        }

        //Check if playing or not playing and handle threading + mediaplayer appropriately
        if(channelRooster.isPlaying()) {
            try {
                mediaPlayer.pause();
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
            //If a previous task exists, kill it
            if(oneInstanceTaskFuture != null) oneInstanceTaskFuture.cancel(true);
            clearChannelRoosterMediaChecks();
            channelRooster.setPaused(true);
            mAdapter.notifyDataSetChanged();
        } else if(channelRooster.isPaused() && mediaPlayer.getCurrentPosition() > 0) {
            clearChannelRoosterMediaChecks();
            try {
                mediaPlayer.seekTo(mediaPlayer.getCurrentPosition());
                mediaPlayer.start();
                channelRooster.setDownloading(false);
                channelRooster.setPlaying(true);
                mAdapter.notifyDataSetChanged();
            } catch (IllegalStateException e) {
                e.printStackTrace();
                channelRooster.setDownloading(false);
                channelRooster.setPlaying(false);
                mAdapter.notifyDataSetChanged();
            }
        } else {
            try {
                mediaPlayer.release();
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
            //If a previous task exists, kill it
            if(oneInstanceTaskFuture != null) oneInstanceTaskFuture.cancel(true);
            clearChannelRoosterMediaChecks();
            mAdapter.notifyDataSetChanged();
            //Start a new task thread
            oneInstanceTaskFuture = executor.submit(new oneInstanceTask());
        }
    }

    private void clearChannelRoosterMediaChecks() {
        //Clear all audio load spinners and set to not playing
        for (ChannelRooster rooster :
                channelRoosters) {
            rooster.setDownloading(false);
            rooster.setPlaying(false);
            rooster.setPaused(false);
        }
    }
}

File: app/src/main/java/com/roostermornings/android/activity/DiscoverFragmentActivity.kt
/*
 * Rooster Mornings Android.
 * Copyright (c)  2017 Roosta Media. All rights reserved.
 */

package com.roostermornings.android.activity

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.os.Bundle
import android.os.RemoteException
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SearchView
import android.view.KeyEvent
import android.view.Menu
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import android.widget.MediaController
import android.widget.TextView
import android.widget.Toast
import butterknife.BindView
import com.roostermornings.android.BaseApplication
import com.roostermornings.android.R
import com.roostermornings.android.activity.base.BaseActivity
import com.roostermornings.android.adapter.DiscoverListAdapter
import com.roostermornings.android.adapter_data.ChannelManager
import com.roostermornings.android.dagger.RoosterApplicationComponent
import com.roostermornings.android.firebase.FA
import com.roostermornings.android.service.MediaService
import com.roostermornings.android.util.JSONPersistence
import com.roostermornings.android.util.RoosterUtils
import com.roostermornings.android.util.Toaster
import io.realm.Realm
import javax.inject.Inject
import kotlin.collections.ArrayList

class DiscoverFragmentActivity : BaseActivity(), DiscoverListAdapter.DiscoverAudioSampleInterface, MediaController.MediaPlayerControl {

    private val mThis = this

    @BindView(R.id.discoverListView)
    lateinit var mRecyclerView: RecyclerView
    @BindView(R.id.toolbar_title)
    lateinit var toolbarTitle: TextView
    @BindView(R.id.button_bar)
    lateinit var buttonBarLayout: LinearLayout
    @BindView(R.id.swiperefresh)
    lateinit var swipeRefreshLayout: SwipeRefreshLayout

    //Set up adapter for monitoring channelrooster media item objects
    private val mediaItems = ArrayList<MediaBrowserCompat.MediaItem>()
    private var mAdapter = DiscoverListAdapter(mediaItems, this@DiscoverFragmentActivity)

    private var mMediaController: MediaControllerCompat? = null
    private lateinit var mMediaBrowser: MediaBrowserCompat

    @Inject
    lateinit var jsonPersistence: JSONPersistence
    @Inject
    lateinit var channelManager: ChannelManager
    @Inject
    lateinit var realm: Realm

    override fun inject(component: RoosterApplicationComponent) {
        component.inject(this)
    }

    private val subscriptionCallback = object : MediaBrowserCompat.SubscriptionCallback() {
        override fun onChildrenLoaded(parentId: String, children: List<MediaBrowserCompat.MediaItem>) {
            super.onChildrenLoaded(parentId, children)
            mediaItems.clear()
            mediaItems.addAll(children)
            mAdapter.notifyDataSetChanged()
            swipeRefreshLayout.isRefreshing = false
        }

        override fun onError(parentId: String) {
            super.onError(parentId)
            swipeRefreshLayout.isRefreshing = false
        }

        override fun onChildrenLoaded(parentId: String, children: List<MediaBrowserCompat.MediaItem>, options: Bundle) {
            super.onChildrenLoaded(parentId, children, options)
            mediaItems.clear()
            mediaItems.addAll(children)
            mAdapter.notifyDataSetChanged()
            swipeRefreshLayout.isRefreshing = false
        }

        override fun onError(parentId: String, options: Bundle) {
            super.onError(parentId, options)
            swipeRefreshLayout.isRefreshing = false
        }
    }

    private val connectionCallback = object : MediaBrowserCompat.ConnectionCallback() {
        override fun onConnected() {
            super.onConnected()

            mMediaBrowser.unsubscribe(mMediaBrowser.root)
            mMediaBrowser.subscribe(mMediaBrowser.root, subscriptionCallback)
            try {
                // Create a MediaControllerCompat
                mMediaController = MediaControllerCompat(mThis, mMediaBrowser.sessionToken)
                mMediaController?.registerCallback(mediaControllerCallback)
                // Save the controller
                MediaControllerCompat.setMediaController(this@DiscoverFragmentActivity, mMediaController)
            } catch (e: RemoteException) {
                e.printStackTrace()
            }
        }

        override fun onConnectionSuspended() {
            super.onConnectionSuspended()
            val mediaController = MediaControllerCompat
                    .getMediaController(mThis)
            if (mediaController != null) {
                mediaController.unregisterCallback(mediaControllerCallback)
                MediaControllerCompat.setMediaController(mThis, null)
            }
        }

    }

    private val mediaControllerCallback = object : MediaControllerCompat.Callback() {

        override fun onPlaybackStateChanged(state: PlaybackStateCompat?) {
            super.onPlaybackStateChanged(state)
            mAdapter.setPlaybackState(state)
            mAdapter.notifyDataSetChanged()
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            super.onMetadataChanged(metadata)
            if (metadata != null) {
                mAdapter.setCurrentMediaMetadata(metadata)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initialize(R.layout.activity_discover)

        BaseApplication.roosterApplicationComponent.inject(this)

        //Notify user of no internet connection
        checkInternetConnection()

        //Final context to be used in threads
        val context = this

        //UI setup thread
        object : Thread() {
            override fun run() {
                //Set highlighting of button bar
                setButtonBarSelection()
                //Set toolbar title
                val toolbar = setupToolbar(toolbarTitle, getString(R.string.discover))
                toolbar?.setNavigationIcon(R.drawable.md_nav_back)
                toolbar?.setNavigationOnClickListener { startHomeActivity() }

                // use this setting to improve performance if you know that changes
                // in content do not change the layout size of the RecyclerView
                mRecyclerView.setHasFixedSize(true)
                //Use a linear layout manager
                val mLayoutManager = LinearLayoutManager(context)
                mRecyclerView.layoutManager = mLayoutManager
                mRecyclerView.adapter = mAdapter

                //Check if uses explore user prop has been set, and if not then create shared pref with default no
                val sharedPrefUsesExplore = sharedPreferences.getString(
                        FA.UserProp.uses_explore.shared_pref_uses_explore, "null")
                if (!sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.no)
                        && !sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.started_explore_playback)
                        && !sharedPrefUsesExplore.contains(FA.UserProp.uses_explore.completed_explore_playback)) {
                    sharedPreferences.edit().putString(FA.UserProp.uses_explore.shared_pref_uses_explore,
                            FA.UserProp.uses_explore.no).apply()
                    FA.SetUserProp(FA.UserProp.uses_explore::class.java, FA.UserProp.uses_explore.no)
                }
            }
        }.run()

//        realm.where(RoosterMediaItem::class.java)
//                .findAll()
//                .takeIf { it.isNotEmpty() }
//                .also {
//            if(checkInternetConnection() && !swipeRefreshLayout.isRefreshing)
//                swipeRefreshLayout.isRefreshing = true
//        }?.let {
//            realmMediaItems ->
//            mediaItems.clear()
//            val gson = Gson()
//            realmMediaItems.forEach {
//                val parcel = gson.fromJson<Parcel>(it.jsonParcel, Parcel::class.java)
//                val mediaItem = MediaBrowserCompat.MediaItem.CREATOR.createFromParcel(parcel)
//                mediaItems.add(mediaItem)
//            }
//            mAdapter.notifyDataSetChanged()
//            swipeRefreshLayout.isRefreshing = false
//        }
        if(checkInternetConnection() && !swipeRefreshLayout.isRefreshing)
            swipeRefreshLayout.isRefreshing = true

        /*
        * Sets up a SwipeRefreshLayout.OnRefreshListener that is invoked when the user
        * performs a swipe-to-refresh gesture.
        */
        swipeRefreshLayout.setOnRefreshListener {
            // This method performs the actual data-refresh operation.
            // The method calls setRefreshing(false) when it's finished.
            refreshData()
        }

        //Set volume rocker to alarm stream
        volumeControlStream = AudioManager.STREAM_MUSIC
    }

    private fun refreshData() {
        if (mMediaBrowser.isConnected) {
            mMediaController?.sendCommand(MediaService.Companion.CustomCommand.REFRESH.toString(), null, null)
        } else {
            swipeRefreshLayout.isRefreshing = false
            if (checkInternetConnection())
                Toaster.makeToast(this, "Failed to refresh, please try again.", Toast.LENGTH_SHORT)
        }
    }

    override fun onStart() {
        super.onStart()
        //Display notifications
        updateRoosterNotification()
        updateRequestNotification()

        //Initialise mediabrowser -- done here so that instance recreated
        //https://stackoverflow.com/questions/35752575/kotlin-lazy-properties-and-values-reset-a-resettable-lazy-delegate
        mMediaBrowser = MediaBrowserCompat(this, ComponentName(this, MediaService::class.java), connectionCallback, null)

        if (!mMediaBrowser.isConnected)
            mMediaBrowser.connect()
    }

    public override fun onPause() {
        super.onPause()

//        //Persist channel roosters for seamless loading
//        if (!mediaItems.isEmpty()) {
//            realm.executeTransaction {
//                realm.where(RoosterMediaItem::class.java).findAll().deleteAllFromRealm()
//                val gson = Gson()
//                mediaItems.forEach {
//                    val roosterMediaItem = RoosterMediaItem()
//                    val parcel = Parcel.obtain()
//                    it.writeToParcel(parcel, 0)
//                    roosterMediaItem.jsonParcel = gson.toJson(parcel)
//                    realm.insert(roosterMediaItem)
//                }
//            }
//        }

        // If media not playing, stop the media service
        //if(mMediaController?.playbackState?.state != PlaybackStateCompat.STATE_PLAYING)
        //    mMediaController?.transportControls?.stop()

        //Unsubscribe and unregister MediaControllerCompat callbacks
        MediaControllerCompat.getMediaController(this@DiscoverFragmentActivity)?.unregisterCallback(mediaControllerCallback)
        if (mMediaBrowser.isConnected) {
            mMediaBrowser.unsubscribe(mMediaBrowser.root, subscriptionCallback)
            mMediaBrowser.disconnect()
        }
    }

    override fun onNewIntent(intent: Intent) {
        handleIntent(intent)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return false
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (RoosterUtils.hasLollipop()) {
            return super.onKeyDown(keyCode, event)
        }
        when (keyCode) {
            KeyEvent.KEYCODE_MEDIA_PLAY -> {
                mMediaController?.dispatchMediaButtonEvent(event)
                return true
            }
            KeyEvent.KEYCODE_MEDIA_PAUSE -> {
                mMediaController?.dispatchMediaButtonEvent(event)
                return true
            }
            KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE -> {
                mMediaController?.dispatchMediaButtonEvent(event)
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            val query = intent.getStringExtra(SearchManager.QUERY)
            //use the query to search your data somehow
            handleSearch(query)
        }
    }

    private fun handleSearch(query: String) {
        mAdapter.refreshAll(mediaItems)
        mAdapter.filter?.filter(query)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_explore, menu)

        // Associate searchable configuration with the SearchView
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.search_explore).actionView as SearchView
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(componentName))

        //When searchView is closed, refresh data
        searchView.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {

            override fun onViewDetachedFromWindow(arg0: View) {
                // search was detached/closed
                mAdapter.refreshAll(mediaItems)
            }

            override fun onViewAttachedToWindow(arg0: View) {
                // search was opened
            }
        })

        return true
    }

    override fun onMediaItemSelected(item: MediaBrowserCompat.MediaItem, isPlaying: Boolean) {
        if (isPlaying) {
            mMediaController?.transportControls?.pause()
        } else {
            FA.Log(FA.Event.explore_channel_rooster_play::class.java,
                    FA.Event.explore_channel_rooster_play.Param.channel_title, item.mediaId)

            mMediaController?.transportControls?.pause()
            mediaItems.indexOfFirst { it.mediaId == item.mediaId }.takeIf { it > -1 }?.let {
                mMediaController?.transportControls?.skipToQueueItem(it.toLong())
                mMediaController?.transportControls?.play()
                // Increment, so that new content received tomorrow
                mediaItems[it].mediaId?.let { channelManager.incrementChannelStoryIteration(it) }
            }
        }
    }

    override fun start() {

    }

    override fun pause() {

    }

    override fun getDuration(): Int {
        return 0
    }

    override fun getCurrentPosition(): Int {
        return 0
    }

    override fun seekTo(i: Int) {

    }

    override fun isPlaying(): Boolean {
        return false
    }

    override fun getBufferPercentage(): Int {
        return 0
    }

    override fun canPause(): Boolean {
        return false
    }

    override fun canSeekBackward(): Boolean {
        return false
    }

    override fun canSeekForward(): Boolean {
        return false
    }

    override fun getAudioSessionId(): Int {
        return 0
    }

    companion object {

        val TAG: String = DiscoverFragmentActivity::class.java.simpleName
    }
}
